### PR TITLE
Do not process rows in aggregations if any of the parameters is NULL

### LINKED ISSUE #41528: Variadic aggregations with nullable parameters might use garbage data
> You have to provide the following information whenever possible.

When the return value of an aggregation cannot be nullable, for example a tuple, the aggregation function might use the undelying / garbage data of a NULL value.

Example on `22.8.5.29`:

Data:
```sql
SELECT
    argMax((n, n), n) AS t,
    toTypeName(t)
FROM
(
    SELECT if((number % 3) = 0, NULL, number) AS n
    FROM numbers(10)
)

Query id: b74a4de2-8738-4434-947f-6945f3c22762

┌─t───────────┬─toTypeName(argMax(tuple(n, n), n))────────┐
│ (NULL,NULL) │ Tuple(Nullable(UInt64), Nullable(UInt64)) │
└─────────────┴───────────────────────────────────────────┘
```

Here I was expecting to get `(8, 8)`. Why didn't we? Let's see the data:

```
SELECT
    (n, n),
    n,
    assumeNotNull(n)
FROM
(
    SELECT if((number % 3) = 0, NULL, number) AS n
    FROM numbers(10)
)

Query id: a3fad631-5c79-47c4-9a63-89258cf17848

┌─tuple(n, n)─┬────n─┬─assumeNotNull(n)─┐
│ (NULL,NULL) │ ᴺᵁᴸᴸ │                0 │
│ (1,1)       │    1 │                1 │
│ (2,2)       │    2 │                2 │
│ (NULL,NULL) │ ᴺᵁᴸᴸ │                3 │
│ (4,4)       │    4 │                4 │
│ (5,5)       │    5 │                5 │
│ (NULL,NULL) │ ᴺᵁᴸᴸ │                6 │
│ (7,7)       │    7 │                7 │
│ (8,8)       │    8 │                8 │
│ (NULL,NULL) │ ᴺᵁᴸᴸ │                9 │
└─────────────┴──────┴──────────────────┘
```

The last value of n is a NULLed 9, which is considered bigger than a non-NULLed 8 (because the null qualifiers are removed and ignored).

OTOH, if we force a cast the value will be default'ed:

```
SELECT
    argMax((n, n), n) AS t,
    toTypeName(t)
FROM
(
    SELECT if((number % 3) = 0, NULL, CAST(number, 'Int32')) AS n
    FROM numbers(10)
)

Query id: 52e778ce-4530-4c58-a0d9-2adcaa9d5f41

┌─t─────┬─toTypeName(argMax(tuple(n, n), n))──────┐
│ (8,8) │ Tuple(Nullable(Int32), Nullable(Int32)) │
└───────┴─────────────────────────────────────────┘

1 row in set. Elapsed: 0.002 sec. 
```

This is only because the NULLed 9 was transformed into a NULLed 0 so the result seems fine:

```
SELECT
    (n, n),
    n,
    assumeNotNull(n)
FROM
(
    SELECT if((number % 3) = 0, NULL, CAST(number, 'Int32')) AS n
    FROM numbers(10)
)

Query id: da7dfc57-a5bb-4f13-a8e7-d52c9b0f9aa5

┌─tuple(n, n)─┬────n─┬─assumeNotNull(n)─┐
│ (NULL,NULL) │ ᴺᵁᴸᴸ │                0 │
│ (1,1)       │    1 │                1 │
│ (2,2)       │    2 │                2 │
│ (NULL,NULL) │ ᴺᵁᴸᴸ │                0 │
│ (4,4)       │    4 │                4 │
│ (5,5)       │    5 │                5 │
│ (NULL,NULL) │ ᴺᵁᴸᴸ │                0 │
│ (7,7)       │    7 │                7 │
│ (8,8)       │    8 │                8 │
│ (NULL,NULL) │ ᴺᵁᴸᴸ │                0 │
└─────────────┴──────┴──────────────────┘

10 rows in set. Elapsed: 0.001 sec.
```

The problem now is that the 0 can then cause issues with the min :D: 

```
Select argMin((n, n), n) t, toTypeName(t) FROM (Select if(number % 3 == 0, NULL, number::Int32) as n from numbers(5, 10))

SELECT
    argMin((n, n), n) AS t,
    toTypeName(t)
FROM
(
    SELECT if((number % 3) = 0, NULL, CAST(number, 'Int32')) AS n
    FROM numbers(5, 10)
)

Query id: 71f7283d-c7db-48cb-91f9-2d1f6a34d790

┌─t───────────┬─toTypeName(argMin(tuple(n, n), n))──────┐
│ (NULL,NULL) │ Tuple(Nullable(Int32), Nullable(Int32)) │
└─────────────┴─────────────────────────────────────────┘

1 row in set. Elapsed: 0.002 sec. 
```

And removing the cast now returns the good value:

```
SELECT
    argMin((n, n), n) AS t,
    toTypeName(t)
FROM
(
    SELECT if((number % 3) = 0, NULL, number) AS n
    FROM numbers(5, 10)
)

Query id: c1292787-23d5-4402-bbdf-5b51e6b8e028

┌─t─────┬─toTypeName(argMin(tuple(n, n), n))────────┐
│ (5,5) │ Tuple(Nullable(UInt64), Nullable(UInt64)) │
└───────┴───────────────────────────────────────────┘

1 row in set. Elapsed: 0.002 sec. 
```

Basically I'm expecting:
* NULL values, in any column, will **always be ignored**; which is similar to what unary aggregation do (or binary+ aggregations that can return NULL).
* If no value was added to the aggregation and the return value is not nullable, then it should return the default value. In this case `(NULL,NULL)::Tuple(Nullable(UInt64), Nullable(UInt64))`

Seems like a straightforward change (but something that could be considered breaking change). Not sure if I'm missing something. Thoughts?

### PR DESCRIPTION
**Pull request by Algunenano**
### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Do not process rows in aggregations if any of the parameters is NULL.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

This was already the case most of the time. It wasn't the case on `windowFunnel`, `intervalLengthSum` or some aggregations with multiple parameters where the return value couldn't be null, for example argMax(tuple, nullable_value).

Tagging as backwards incompatible change because it's possible (though unlikely) that some people were relying on this behaviour. Note that if that was the case that was equivalent to expect `assumeNotNull(x)` to be the default value, which is absolutely wrong. Instead recommend using `COALESCE` to set up a proper value in case of NULL, which depends on the use case.

Closes https://github.com/ClickHouse/ClickHouse/issues/41528

### CODE REVIEW

**Review by alexey-milovidov — APPROVED** [review]

**Comment by filimonov** [comment]
Wondering why it's not in the changelog.

**Comment by Algunenano** [comment]
> Wondering why it's not in the changelog.

I guess it was bad timing, since the [changelog PR](https://github.com/ClickHouse/ClickHouse/pull/41668) and this one got merge with a couple of hours of difference.

File: src/AggregateFunctions/AggregateFunctionNull.cpp [hunk 1]
```diff
@@ -108,14 +108,14 @@ class AggregateFunctionCombinatorNull final : public IAggregateFunctionCombinato
...
-                return std::make_shared<AggregateFunctionNullVariadic<true, true, true>>(nested_function, arguments, params);
+                return std::make_shared<AggregateFunctionNullVariadic<true, true>>(nested_function, arguments, params);
             }
             else
             {
                 if (serialize_flag)
-                    return std::make_shared<AggregateFunctionNullVariadic<false, true, true>>(nested_function, arguments, params);
+                    return std::make_shared<AggregateFunctionNullVariadic<false, true>>(nested_function, arguments, params);
                 else
-                    return std::make_shared<AggregateFunctionNullVariadic<false, true, false>>(nested_function, arguments, params);
```
Comment (Alex-Cheng) [inline src/AggregateFunctions/AggregateFunctionNull.cpp:118]: I am just wondering why the second template argument `serialize_flag` for `AggregateFunctionNullVariadic` is set to true when the local variable `serialize_flag` is **false**.

Comment (Algunenano) [reply src/AggregateFunctions/AggregateFunctionNull.cpp:118]: It’s a bug that already existed before the change. Unfortunately fixing it is a breaking change and not worth it. There is a comment in the code now: https://github.com/ClickHouse/ClickHouse/blob/f2bad692e2835c58807daba0d4fa21eb4f440c1d/src/AggregateFunctions/Combinators/AggregateFunctionNull.cpp#L161

**Pull request merged**